### PR TITLE
refactor: rewrite services, add function to check if a resource is ready …

### PR DESCRIPTION
**Pull request by cristiGuranIonos**
…or deleted

## What does this fix or implement?

Services were not really useful. Rewrite services so they are the place where processing happens. Remove useless interfaces.
Rename container-registry package to have a proper name.
Add generic waitfor functions to check for resource availability, or deletion.
ApiResponseInfo - interface used to abstract apiresponse types from different sdks.

## Checklist

- [ ] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [X] Tests added or updated
- [ ] Documentation updated
- [X] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

File: utils/utils.go [hunk 1]
```diff
@@ -205,3 +208,47 @@ func RemoveNewLines(s string) string {
...
+type ApiResponseInfo interface {
+	HttpNotFound() bool
+	LogInfo()
+}
+
+// ResourceReadyFunc polls api to see if resource exists based on id
+type ResourceReadyFunc func(ctx context.Context, d *schema.ResourceData) (bool, error)
+
+// WaitForResourceToBeReady - keeps retrying until resource is ready(true is returned), or until err is thrown, or ctx is cancelled
+func WaitForResourceToBeReady(ctx context.Context, d *schema.ResourceData, fn ResourceReadyFunc) error {
```
Comment (adeatcu-ionos) [inline utils/utils.go:222]: @cristiGuranIonos I noticed that this function is only used for certificate manager. In this PR we also have changes for mongo clusters, users, etc. Did you forget to also use `WaitForResourceToBeReady` for those resources or you just added this as a starting point for future PRs?

**Review by adeatcu-ionos — CHANGES_REQUESTED** [review]

**Comment by adeatcu-ionos** [comment]
@cristiGuranIonos I'm not familiar with the `wait for response` logic in `golang` but the code looked all right. Also, I'm not sure if there could be any improvements related to the design/the way in which you modified the services logic, at this point I can't write pertinent comments about this, but overall the code looks more cleaner now, so good job!

**Comment by cristiGuranIonos** [comment]
> @cristiGuranIonos I'm not familiar with the `wait for response` logic in `golang` but the code looked all right. Also, I'm not sure if there could be any improvements related to the design/the way in which you modified the services logic, at this point I can't write pertinent comments about this, but overall the code looks more cleaner now, so good job!

Uses channels, has added complexity, also does not have support for nested resources.  It's also only usable for ionoscloud due to the way the interface has been defined there.

File: utils/utils.go (hunk 1 above)
Comment (cristiGuranIonos) [reply utils/utils.go:222]: Exactly, starting point I've implemented it for dataplatform also in the dataplatform PR. The others will do in separate PRs so less code.

File: services/dbaas/users.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+func (c *MongoClient) GetUsers(ctx context.Context, clusterId string) (mongo.UsersList, utils.ApiResponseInfo, error) {
+	users, apiResponse, err := c.sdkClient.UsersApi.ClustersUsersGet(ctx, clusterId).Execute()
+	apiResponse.LogInfo()
+	return users, apiResponse, err
+}
+
+func (c *MongoClient) FindUserByUsername(ctx context.Context, clusterId, username string) (mongo.User, utils.ApiResponseInfo, error) {
+	user, apiResponse, err := c.sdkClient.UsersApi.ClustersUsersFindById(ctx, clusterId, DefaultMongoDatabase, username).Execute()
+	apiResponse.LogInfo()
```
[the remark this answers is not in the document]
Comment (cristiGuranIonos) [reply services/dbaas/users.go:29]: The function works without checking for resp to be null. It does it inside the function, so it's not necessary to check here.

File: ionoscloud/resource_dbaas_mongodb_user.go [hunk 3]
```diff
@@ -159,13 +155,13 @@ func resourceDbaasMongoUserUpdate(ctx context.Context, d *schema.ResourceData, m
 		request.Properties.Roles = &roles
 	}
 
-	_, _, err := client.UsersApi.ClustersUsersPatch(ctx, clusterId, defaultMongoDatabase, username).PatchUserRequest(request).Execute()
+	_, err := client.UpdateUser(ctx, clusterId, username, request)
```
[the remark this answers is not in the document]
Comment (cristiGuranIonos) [reply ionoscloud/resource_dbaas_mongodb_user.go:158]: For mongo it's actually in the process of being removed for good, so it won't be used at all.

File: services/dbaas/cluster.go [hunk 4]
```diff
@@ -11,24 +11,8 @@ import (
...
-	ListClusters(ctx context.Context, filterName string) (mongo.ClusterList, *mongo.APIResponse, error)
-	CreateCluster(ctx context.Context, cluster mongo.CreateClusterRequest) (mongo.ClusterResponse, *mongo.APIResponse, error)
-	//UpdateCluster(ctx context.Context, clusterId string, cluster psql.PatchClusterRequest) (psql.ClusterResponse, *psql.APIResponse, error)
-	DeleteCluster(ctx context.Context, clusterId string) (mongo.ClusterResponse, *mongo.APIResponse, error)
-}
-
 func (c *PsqlClient) GetCluster(ctx context.Context, clusterId string) (psql.ClusterResponse, *psql.APIResponse, error) {
-	cluster, apiResponse, err := c.ClustersApi.ClustersFindById(ctx, clusterId).Execute()
+	cluster, apiResponse, err := c.sdkClient.ClustersApi.ClustersFindById(ctx, clusterId).Execute()
 	if apiResponse != nil {
```
[the remark this answers is not in the document]
Comment (cristiGuranIonos) [reply services/dbaas/cluster.go:16]: Correct, will do when I finish refactoring dbaas in a future PR. There are some other things to change in the service functions.

**Pull request merged**